### PR TITLE
cephadm: correct ipv6 support in port open detection

### PR DESCRIPTION
**Pull request by pcuzner**
Updated the port_in_use function to work better in ipv4 and ipv6 environments. 

Signed-off-by: Paul Cuzner <pcuzner@redhat.com>

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -108,13 +108,25 @@ def port_in_use(port_num):
     # type (int) -> bool
     """Detect whether a port is in use on the local machine - IPv4 and IPv6"""
 
+    def attempt_bind(address):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]:
```suggestion
    def attempt_bind(s, address):
```

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -108,13 +108,25 @@ def port_in_use(port_num):
...
+        # type (str) -> None
+        try:
+            s.bind((address, port_num))
+        except (socket.error, OSError) as e:  # py2 and py3
+            if e.errno == 98:                 # Address already in use
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]:
```suggestion
            if e.errno == errno.EADDRINUSE:
```
?

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -108,13 +108,25 @@ def port_in_use(port_num):
...
+                raise OSError
+            elif e.errno == 99:               # Cannot assign requested address
```
Comment (mgfritch) [inline src/cephadm/cephadm]:
```suggestion
            elif e.errno == errno.EADDRNOTAVAIL:               # Cannot assign requested address
```

File: src/cephadm/cephadm [hunk 4]
```diff
@@ -108,13 +108,25 @@ def port_in_use(port_num):
...
+        attempt_bind('0.0.0.0')
         s.close()
+
         s = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
-        s.bind(("::1", port_num))
+        attempt_bind('::/0')
         s.close()
+
     except OSError:
         s.close()
```
Comment (mgfritch) [inline src/cephadm/cephadm]:
```suggestion
    finally:
        s.close()
```

Comment (mgfritch) [reply src/cephadm/cephadm]: Close the socket in one place via `finally` block ?

File: src/cephadm/cephadm [hunk 5]
```diff
@@ -108,13 +108,25 @@ def port_in_use(port_num):
...
+            s.bind((address, port_num))
+        except (socket.error, OSError) as e:  # py2 and py3
+            if e.errno == 98:                 # Address already in use
+                raise OSError
+            elif e.errno == 99:               # Cannot assign requested address
+                return
+
     try:
         s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        s.bind(("127.0.0.1", port_num))
```
Comment (mgfritch) [inline src/cephadm/cephadm]: Add arg to specify the ip addr during deploy? 

e.g. `cephadm deploy --mon-ip 0.0.0.0` or similar?

File: src/cephadm/cephadm (hunk 2 above)
Comment (pcuzner) [reply src/cephadm/cephadm]: ACK - done

File: src/cephadm/cephadm (hunk 3 above)
Comment (pcuzner) [reply src/cephadm/cephadm]: Done

File: src/cephadm/cephadm (hunk 4 above)
Comment (pcuzner) [reply src/cephadm/cephadm]: Done. close moved to inner function's finally block

File: src/cephadm/cephadm (hunk 5 above)
Comment (pcuzner) [reply src/cephadm/cephadm]: possibly - but I think on a deploy we should validate the IP exists on the host AND the port(s) are free - which is bigger than the scope I was intending for this PR

File: src/cephadm/cephadm [hunk 6]
```diff
@@ -108,15 +109,26 @@ def port_in_use(port_num):
...
-        s.close()
+        attempt_bind('0.0.0.0')
+
         s = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
-        s.bind(("::1", port_num))
-        s.close()
+        attempt_bind('::/0')
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: You're sure? I'm getting an error when trying to bind 8081:

```
In [6]: s = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)

In [7]: s.bind(('::/0', 8081))
---------------------------------------------------------------------------
gaierror                                  Traceback (most recent call last)
<ipython-input-7-1935d55732aa> in <module>()
----> 1 s.bind(('::/0', 8081))

gaierror: [Errno -3] Temporary failure in name resolution
```

This works however:
```
In [8]: s.bind(('::', 8081))

In [9]: 
```

Comment (pcuzner) [reply src/cephadm/cephadm]: thanks for catching!

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Pull request merged**